File: codebase/src/schemas.ts
import { z } from "zod";

export const SessionState = z.enum([
  "attempt_1_open",
  "evaluating_attempt",
  "diagnosis",
  "retry",
  "explain_back",
  "transfer_check",
  "source_review",
  "clarify",
  "out_of_scope",
  "completed"
]);
export type SessionState = z.infer<typeof SessionState>;

export const Confidence = z.enum(["Chắc", "Khá chắc", "Chưa chắc"]);
export const Basis = z.enum(["Đã học trước đó", "Suy luận", "Đoán", "Chưa có căn cứ"]);

export const CreateSessionBody = z.object({
  itemId: z.literal("tokenization-01"),
  itemVersion: z.literal("2026-09-16.1").optional(),
  mode: z.literal("demo").optional()
}).strict();
export type CreateSessionBody = z.infer<typeof CreateSessionBody>;

export const AttemptBody = z.object({
  stateVersion: z.number().int().positive(),
  kind: z.enum(["attempt_1", "retry"]),
  answer: z.object({
    text: z.string().max(500),
    explanation: z.string().max(500)
  }).strict(),
  confidence: Confidence.optional(),
  basis: Basis.optional()
}).strict();
export type AttemptBody = z.infer<typeof AttemptBody>;

export const HintBody = z.object({
  stateVersion: z.number().int().positive(),
  level: z.union([z.literal(1), z.literal(2), z.literal(3)])
}).strict();
export type HintBody = z.infer<typeof HintBody>;

export const ExplainBody = z.object({
  stateVersion: z.number().int().positive(),
  text: z.string().min(1).max(500)
}).strict();
export type ExplainBody = z.infer<typeof ExplainBody>;

export const TransferBody = z.object({
  stateVersion: z.number().int().positive(),
  answer: z.string().max(500),
  reasoning: z.string().max(500)
}).strict();
export type TransferBody = z.infer<typeof TransferBody>;

export const StateVersionBody = z.object({
  stateVersion: z.number().int().positive()
}).strict();

export const CoachDraft = z.object({
  action: z.enum(["diagnose_and_hint", "ask_clarification", "abstain"]),
  diagnosisCode: z.string().nullable(),
  confidence: z.enum(["high", "medium", "low"]),
  hintLevel: z.union([z.literal(1), z.literal(2), z.literal(3)]).nullable(),
  citationIds: z.array(z.string()).max(2),
  learnerMessage: z.string().min(1).max(500)
}).strict();
export type CoachDraft = z.infer<typeof CoachDraft>;

export const ExplainClaims = z.object({
  tokenIsNotWordOrLetter: z.boolean(),
  tokenizerCanSplitDifferently: z.boolean(),
  exactCountNeedsTool: z.boolean()
});

File: codebase/backend/app/schemas.py
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field, field_validator
from .sections import SECTIONS

Confidence = Literal["Chắc", "Khá chắc", "Chưa chắc"]
Basis = Literal["Đã học trước đó", "Suy luận", "Đoán", "Chưa có căn cứ"]
SessionState = Literal[
    "attempt_1_open", "evaluating_attempt", "diagnosis", "retry", "explain_back",
    "transfer_check", "source_review", "clarify", "out_of_scope", "completed",
]
SECTION_IDS = {item["sectionId"] for item in SECTIONS}
ITEM_IDS = {item["itemId"] for item in SECTIONS}


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class CreateSessionBody(StrictModel):
    itemId: str | None = None
    itemVersion: str | None = None
    sectionId: str = "prompt-fundamentals"
    mode: Literal["demo"] | None = None

    @field_validator("sectionId")
    @classmethod
    def validate_section(cls, value: str) -> str:
        if value not in SECTION_IDS:
            raise ValueError("SECTION_NOT_FOUND")
        return value

    @field_validator("itemId")
    @classmethod
    def validate_item(cls, value: str | None) -> str | None:
        if value is None:
            return value
        if value not in ITEM_IDS and value != "prompt-clarity-01":
            raise ValueError("ITEM_NOT_FOUND")
        return value


class Answer(StrictModel):
    text: str = Field(max_length=500)
    explanation: str = Field(max_length=500)


class AttemptBody(StrictModel):
    stateVersion: int = Field(gt=0)
    kind: Literal["attempt_1", "retry"]
    answer: Answer
    confidence: Confidence | None = None
    basis: Basis | None = None


class HintBody(StrictModel):
    stateVersion: int = Field(gt=0)
    level: Literal[1, 2, 3]


class ExplainBody(StrictModel):
    stateVersion: int = Field(gt=0)
    text: str = Field(min_length=1, max_length=500)


class TransferBody(StrictModel):
    stateVersion: int = Field(gt=0)
    answer: str = Field(max_length=500)
    reasoning: str = Field(max_length=500)


class StateVersionBody(StrictModel):
    stateVersion: int = Field(gt=0)


class CoachDraft(StrictModel):
    action: Literal["diagnose_and_hint", "ask_clarification", "abstain"]
    diagnosisCode: str | None
    confidence: Literal["high", "medium", "low"]
    hintLevel: Literal[1, 2, 3] | None
    citationIds: list[str] = Field(max_length=2)
    learnerMessage: str = Field(min_length=1, max_length=500)
